Add docs Array.groupByToMap

hamishwillee opened the pull request:
Array grouping proposal added to FF nightly in FF98. 
This adds `groupByToMap` docs. There is an associated interactive example in https://github.com/mdn/interactive-examples/pull/2015.

Additionally:
- Updates to `Array.groupBy()`, following on from https://github.com/mdn/content/pull/12980
- Updated `Array` with links to `groupBy` and `groupByToMap`
- Updated `Array` with grouping example section.
- Added methods to Experimental features page.

Part of #12579

hamishwillee committed:
Add docs Array.groupByToMap

SebastianSimon commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,158 @@
+
+## Description
+
+The `groupBytoMap()` method executes the `callbackFn` function once for each index of the array, returning an object indicating the group of the associated element.
```
`groupBytoMap()` → `groupByToMap()`.

SebastianSimon commented:
The return value of the `groupByToMap` callback isn’t required to be an object; it can be any arbitrary value (same as a `Map` key). Maybe reword “using the _objects_ returned” to “using the arbitrary values returned”? Similarly, “unique objects” → “unique values”? This appears in a few places in this article.

This part already explains that the method is most useful when using objects as keys; I think every remaining instance of “object” can be reworded as “value”:

> This method is useful when mapping related information to a specific object that might change over time. If the mapping object is invariant then you could instead use `Array.prototype.groupBy()` to group the elements, representing the object with a string.

The same nitpick applies to the `groupBy` page which says “If you need to group elements using a key that is some arbitrary object, use `Array.prototype.groupByToMap()` instead” which should also say “arbitrary value” instead.

Another thing I noticed: should this article mention that key equality is based on the [SameValue](https://tc39.es/ecma262/#sec-samevalue) abstract operation (but with `-0` converted to `0`)? The [`includes` article](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes#return_value) mentions it and it may avoid questions in the future that ask why `[ NaN, NaN ].groupByToMap((value) => value).size` is `1`.

Other than that, great article, especially the example using destructuring and explaining how useful it is in this situation.

hamishwillee committed:
Fix up groupByToMap typo. Add experimental tags

hamishwillee committed:
groupByToMap returns values not objects

hamishwillee committed:
Array: Add grouping example section

hamishwillee commented:
@SebastianSimon Thanks for your review - really helpful.
- Updated to use value rather than object. "value" is a spec term rather than a common English or Javascript term, so I have clarified this in the parameters section
  > The value (`object` or `primitive`) returned ...
- Fixed all the typos on the method name - got to love copy/paste :-(
- A few other updates - e.g. `Array` page example of grouping.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -446,6 +447,41 @@ console.log(fruitsAlias);
+Note that the returned object references the _same_ elements as the original array (not {{glossary("deep copy","deep copies")}}).
+Changing the internal structure of these elements will be reflected in both the original array and the returned object.
+
+If you can't use a string as the key, for example, if the information to group is associated with an object that might change, then you can instead use {{jsxref("Array.prototype.groupByToMap()")}}.
```
Note, I only mention `groupByToMap` rather than fully document it. I think the vast majority of use cases will use `groupBy`. People can still follow the link to find more informaiton.

I also haven't mentioned that you could do the same thing with `Array.reduce()`, albeit a lot less succinctly.

hamishwillee committed:
Add to experimental features

hamishwillee commented:
@wbamberg @sideshowbarker Can I please get a review of this?

wbamberg commented:
@hamishwillee I will take a look.

hamishwillee commented:
Thanks!!

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+
+<!-- {{EmbedInteractiveExample("pages/js/array-groupbytomap.html")}} -->
+
+The elements in the {{jsxref("Map")}} and the original array are the same (not {{glossary("deep copy","deep copies")}}).
+Changing the internal structure of the elements will be reflected in both the original array and the returned {{jsxref("Map")}}.
+
+An object that needs to access the groups must keep a reference to the original key; the properties of a key can be modified, but you can't use another object that just happens to have the same name and properties.
+
+This method is useful when mapping related information to a specific object that might change over time.
+If the mapping object is invariant then you could instead use {{jsxref("Array.prototype.groupBy()")}} to group the elements, representing the object with a string.
```
I think this or a version of it should be in the H2#Description section, not here.

Method pages should start with a short paragraph, then interactive example, then H2#Syntax, then any more prose in H2#Description. Like, say, [`filter()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) or [`concat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat) or [`copyWithin()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/copyWithin).

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+The elements in the {{jsxref("Map")}} and the original array are the same (not {{glossary("deep copy","deep copies")}}).
+Changing the internal structure of the elements will be reflected in both the original array and the returned {{jsxref("Map")}}.
+
+An object that needs to access the groups must keep a reference to the original key; the properties of a key can be modified, but you can't use another object that just happens to have the same name and properties.
```
I didn't understand this at all until I read the example further down the page. I expect it's basically impossible to summarise this, so it might be better not to say it until you have space to explain it properly with an example.

SebastianSimon replied:
Is it even necessary to explain any of these object reference behaviors here? This essentially explains how a `Map` works with objects as keys, so I’d rather expect to find this description on the docs page for `Map` instead.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+
+An object that needs to access the groups must keep a reference to the original key; the properties of a key can be modified, but you can't use another object that just happens to have the same name and properties.
+
+This method is useful when mapping related information to a specific object that might change over time.
```
Is it worth adding something like "...because you can modify properties of the key object and still use it as a key to retrieve its group." (assuming I've understood the point of this)?

I guess this is related also to why you might use this instead of `groupBy()`? It would be really nice to have something in this page explicitly addressing that, just to be really clear.

hamishwillee replied:
Yes, that's pretty much the intent, as I understand it from https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c16

I think that we do need a clear statement of why you use this, and I agree with @SebastianSimon probably most of the depth of that should be in the Map docs. I'll remove/restructure as appropriate.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+
+    - `element`
+      - : The current element in the array.
+    - `index` {{optional_inline}}
+      - : The index (position) of the current element in the array.
+    - `array` {{optional_inline}}
```
In both these cases I don't think `Optional` is appropriate here. `Optional` is for arguments you don't have to _pass_, but here the arguments are always passed, and we are using it to suggest that the function doesn't have to _use_ them. But no function ever has to use any of its arguments. - we might as well say `element` is optional in this sense.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+
+## Description
+
+The `groupByToMap()` method executes the `callbackFn` function once for each index of the array, returning a value indicating the group of the associated element.
```
This implies to me that "The `groupByToMap()` method" is the thing that returns "a value indicating the group of the associated element". Perhaps:

```suggestion
The `groupByToMap()` method executes the callback function once for each index of the array. The callback function returns a value indicating the group of the associated element.
```

? (also I think maybe we should consistently use either "the callback function" or "`callbackFn`" in this section)

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+If a `thisArg` parameter is provided to `groupByToMap()`, it will be used as the `this` value inside each invocation of the `callbackFn`.
+If it is not provided, then {{jsxref("undefined")}} is used.
+
+The `groupByToMap()` method does not mutate the array on which it is called, but the function provided to `callbackFn` can.
+Note however that the elements processed by `groupByToMap()` are set _before_ the first invocation of `callbackFn`.
+Therefore:
+
+- `callbackFn` will not visit any elements added to the array after the call to `groupByToMap()` begins.
+- Elements that are assigned to indexes already visited, or to indexes outside the range, will not be visited by `callbackFn`.
+- If an existing, yet-unvisited element of the array is changed by `callbackFn`, its value passed to the `callbackFn` will be the value at the time `groupByToMap()` visits that element's index.
+- Elements that are {{jsxref("Operators/delete", "deleted")}} are still visited.
+
+> **Warning:** Concurrent modifications of the kind described above frequently lead to hard-to-understand code and are generally to be avoided (except in special cases).
```
You might consider using a subheading for this (especially if you migrate some stuff into H2#Description from the earlier intro prose). Like "Mutating the array in the callback" or something. Partly because this is kind of edge-casey, so putting it in a separate section makes it easier for people to ignore.

wbamberg replied:
Not sure if this got missed or you decided not to (I'm fine if you decided not to).

hamishwillee replied:
Hi @wbamberg,
Sorry, I did miss this. It is a very good suggestion. Fixed in both docs.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+## Examples
+
+First we define an array containing objects representing an inventory of different foodstuffs.
+Each food has a `type`, which might need to be stored differently, and a `quantity` that can be used to determine when we need to reorder each item.
```
I don't understand what "which might need to be stored differently" means here. It suggests data storage perhaps when I guess you mean food storage? I think it would be OK to delete this.

SebastianSimon replied:
It’s probably “storing the foodstuffs differently, as a real world application”, i.e. meats should be stored separately in a freezer, etc., but the sentence is indeed confusing, especially because it’s unclear which meaning the word “stored” has (_computer memory_ or _kitchen pantry_?); perhaps “reorder” has the same problem (_sorting_ or _restocking_?). Maybe this:

> First we define an array containing objects representing an inventory of different foodstuffs. Each food item has a `type` and a `quantity`. The `type` may indicate how a given food item might need to be stored in a kitchen pantry, and the `quantity` may be used to determine when we need to restock each item.

Maybe someone can come up with a better example usage for “store”. Maybe an explicit example helps, e.g. _“For example, `{ name: "fish", type: "meat", quantity: 22 }` means that our kitchen has 22 fish items in stock, and, since it’s meat, they need to be stored in the freezer.”_ or something like that. Maybe we can come up with a completely different example that doesn’t involve food items — and how to store them — at all.

The `groupBy` article has exactly the same example and should be edited as well.

hamishwillee replied:
Thanks. Yes this is a problem. I will fix it. Probably using text like Sebastians.

wbamberg replied:
> It’s probably “storing the foodstuffs differently, as a real world application”

That was my guess, but I didn't want to assume :). I agree that removing the line is better than trying to explain it more.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+- `callbackFn` will not visit any elements added to the array after the call to `groupByToMap()` begins.
+- Elements that are assigned to indexes already visited, or to indexes outside the range, will not be visited by `callbackFn`.
+- If an existing, yet-unvisited element of the array is changed by `callbackFn`, its value passed to the `callbackFn` will be the value at the time `groupByToMap()` visits that element's index.
+- Elements that are {{jsxref("Operators/delete", "deleted")}} are still visited.
```
```suggestion
- Elements that are {{jsxref("Operators/delete", "deleted", "nocode")}} are still visited.
```

...since "deleted" isn't a keyword it should not get syntax highlighting. But also, is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete correct here? That's not how you delete an element from an array, is it?

hamishwillee replied:
Thanks! Yes, that is how you delete elements from an array, leaving an empty index for the element you deleted. 
Since of course you might more naturally use a method that returns a new array with specific elements remove (e.g. `splice()` I've updated the link to make this more clear.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
````diff
@@ -0,0 +1,160 @@
+The returned `result` object is a `Map` so we need to call `get()` with the key to obtain the array.
+
+```js
+let reorder  = { reorder: true };
````
```suggestion
const reorder  = { reorder: true };
```

I'm Mr Let-to-const. Although you do `reorder['fast']  = true ;`, that's allowed because `reorder` is not reassigned.

wbamberg replied:
Also it's a bit confusing that the object variable is `reorder`, but it contains a boolean also called `reorder`. I don't have a great idea about how to improve this though.

SebastianSimon replied:
Well, honestly, the example is strange anyway. Why would you need objects as keys here? What is the motivation behind needing to keep track of references for the keys? In which scenario is mutating the key a good idea? I don’t currently have a better alternative, either way. Just a small note: if we reword “reorder” to “restock”, this identifier needs to be updated here as well.

hamishwillee replied:
Thanks. I made the change to restock here and everywhere. 
Yes, the example is bogus, but I suspect there is no "simple" real world simple example where you would do this. If you look at the developer explanation in https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c16 it seems clear to me that it certainly isn't something you would need to do every day :-)

If you look at line 81 in updated doc I hope the explanation of why at least you might use this makes more sense, even though no intuitive example is provided.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
````diff
@@ -0,0 +1,160 @@
+This is a very succinct way to access the relevant values of elements within a function.
+
+
+An object key can be modified and still used.
+However you can't recreate the key and still use it.
+For this reason it is important that anything that needs to use the map keeps a reference to its keys.
+
+```js
+// A modified key can still be used in a Map
+reorder['fast']  = true ;
+console.log(result.get(reorder));
+// expected output: Array [Object { name: "bananas", type: "fruit", quantity: 5 }]
+
+// But you can't use a different key, even if it has the same structure!
+const reorder2  = { reorder: true };
+console.log(result.get(reorder2));
+// expected output: undefined
+```
````
Yeah, this is a very nice explanation of the confusing paragraph at the start.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
````diff
@@ -0,0 +1,160 @@
+// expected output: undefined
+```
+
+The callback syntax provides access to the array and current index, so it is possible to implement grouping strategies based on the values of other elements in the array.
````
This is interesting, and might be better in H2#Description than in H2#Examples.

hamishwillee replied:
I had it here to indicate "this is a simple example - you can do a bit more if you like". As you say, probably better in the description. Text added

> `callbackFn` is called with the value of the current element, the current index, and the array itself.
While groups often depend only on the current element, it is possible to implement grouping strategies based on the values of other elements in the array.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,160 @@
+## See also
+
+- {{jsxref("Array.prototype.groupBy()")}}
+- [Polyfill of `Array.prototype.groupBy` in `core-js`](https://github.com/zloirock/core-js#array-grouping)
```
```suggestion
- [Polyfill of `Array.prototype.groupByToMap()` in `core-js`](https://github.com/zloirock/core-js#array-grouping)
```
?

SebastianSimon replied:
Are the parentheses actually needed?

teoli2003 replied:
Yes, we put parenthesis after method names to distinguish the from properties/attributes.

wbamberg replied:
It seems like this change is not applied?

wbamberg reviewed (changes requested):
Thanks @hamishwillee ! I had a few more or less vague comments.

hamishwillee committed:
Callback params not options

hamishwillee committed:
Update files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Example - remove type/quantity justifications

hamishwillee committed:
Update files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Merge branch 'ff98arraybytomap' of https://github.com/hamishwillee/content into ff98arraybytomap

hamishwillee committed:
Move the note about more complex examples into the description

hamishwillee committed:
Improve on when to use this method

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,165 @@
+
+<!-- {{EmbedInteractiveExample("pages/js/array-groupbytomap.html")}} -->
+
+The method is primarily useful when grouping elements that are associated with an object, and in particular when that object might change over time.
```
@wbamberg I've moved most of the info out of the overview down to the description - here just a summary and alternative.

hamishwillee commented:
Thanks @wbamberg @SebastianSimon 

Hopefully better. Ready for you guys to have another look.

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,165 @@
+This means it may be less efficient for sparse arrays, compared to methods that only visit assigned values.
+
+If a `thisArg` parameter is provided to `groupByToMap()`, it will be used as the `this` value inside each invocation of the `callbackFn`.
+If it is not provided, then {{jsxref("undefined")}} is used.
```
This does not appear to be correct. Trying our the example, with extra logging:

```js
const inventory = [
  { name: 'asparagus', type: 'vegetables', quantity: 9 },
  { name: 'bananas',  type: 'fruit', quantity: 5 },
  { name: 'goat', type: 'meat', quantity: 23 },
  { name: 'cherries', type: 'fruit', quantity: 12 },
  { name: 'fish', type: 'meat', quantity: 22 }
];

let restock  = { restock: true };
const sufficient = { restock: false };
const result = inventory.groupByToMap( ({ quantity }) => {
  console.log(`it is: ${this}`)
  return quantity < 6 ? restock : sufficient
});
console.log(result.get(restock));
```

...I get:

```
it is: [object Window]
```
```

hamishwillee replied:
@wbamberg Interesting! This is what the [spec says](https://tc39.es/proposal-array-grouping/#sec-array.prototype.groupbymap), so I propose that we merge this.

I will of course still chase this in https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c17

wbamberg replied:
Yes, I agree we should document the spec, and thanks for pursuing this in the bug!

hamishwillee replied:
Finally https://github.com/mdn/content/pull/14421

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,165 @@
+    - `index`
+      - : The index (position) of the current element in the array.
+    - `array`
+      - : The array that `groupBy()` was called on.
```
```suggestion
      - : The array that `groupByToMap()` was called on.
```

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
```diff
@@ -0,0 +1,165 @@
+This is because even if the object is modified, it will continue to work as a key to the returned `Map`.
+If you instead create a string representation for the object and use that as a grouping key in {{jsxref("Array.prototype.groupBy()")}}, you must maintain the mapping between the original object and its representation as the object changes.
+
+> **Note:** An object that needs to access the groups must keep a reference to the _original_ key (although you may modify its properties).
```
I think "An object" here is confusing, since we are talking about the objects used as keys already, but this is just "the caller". Maybe something like: 

```suggestion
> **Note:** To access the groups in the returned `Map`, you must use the same object that was originally used as a key in the `Map` (although you may modify its properties).
```

hamishwillee replied:
Yes. Thanks!

wbamberg commented on files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md:
````diff
@@ -0,0 +1,165 @@
+The returned `result` object is a `Map` so we need to call `get()` with the key to obtain the array.
+
+```js
+let restock  = { restock: true };
````
This is still `let` and should be `const`, per https://github.com/mdn/content/pull/13171#discussion_r814467008.

wbamberg reviewed (changes requested):
Thanks for the updates Hamish! There were a couple of things that didn't get applied. I'm happy to discuss if they are not needed but it would be good at least to agree on that.

hamishwillee committed:
Add Mutating the array in the callback section

hamishwillee committed:
fix a let to const case

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented:
Thanks very much @wbamberg - appreciate the care you're taking. 
I've made the fixes as suggested. There is one open, but I think we can merge and I will follow up separately: https://github.com/mdn/content/pull/13171#discussion_r816179211

Ready for another look.

wbamberg reviewed (approved):
👍 thanks for the updates Hamish!
